### PR TITLE
Feat: add --datapath flag for custom data folder

### LINKED ISSUE #401: 能否添加命令行选项并且不需依赖 gopath?
如题

### PR DESCRIPTION
**Pull request by Loyalsoldier**
Fixes #401 

- add `--datapath` flag to use custom data folder to generate `dlc.dat`
- If no `--datapath` flag is set, use `$GOPATH` instead

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -139,7 +142,21 @@ func parseEntry(line string) (Entry, error) {
...
+			return path, nil
+		}
+
+		if isFlagSet {
+			err = errors.New("The path you input using flag '--datapath' does not exist")
+		} else if !isEnvSet {
+			err = errors.New("Folder 'data' in the same directory as 'main.go' does not exist")
+		}
+		return "", err
+	}
```
Comment (iochen) [inline main.go:158]:
```
func DetectPath(path string) (string, error) {
	arrPath := strings.Split(path, string(filepath.ListSeparator))
	for _, content := range arrPath {
		fullPath := filepath.Join(content, "src", "github.com", "v2ray", "domain-list-community", "data")
		_, err := os.Stat(fullPath)
		if err == nil || os.IsExist(err) {
			return fullPath, nil
		}
	}
	err := errors.New("no file found")
	return "", err
}
```

File: main.go [hunk 2]
```diff
@@ -215,8 +232,39 @@ func ParseList(list *List, ref map[string]*List) (*ParsedList, error) {
...
+	var dir string
+	var err error
+	if isSetDataPath {
+		dir, err = DetectPath(*dataPath, isSetDataPath, isSetGOPATH)
+	} else if isSetGOPATH {
+		dir, err = DetectPath(os.Getenv("GOPATH"), isSetDataPath, isSetGOPATH)
+	} else {
+		fmt.Println("Notice: neither flag '--datapath' nor $GOPATH is set, using './data' folder by default.")
+		dir, err = DetectPath("./data", isSetDataPath, isSetGOPATH)
+	}
```
Comment (iochen) [inline main.go:266]:
```
var dir string
var err error
if *dataPath != "" {
	dir, err = DetectPath(*dataPath)
} else {
	dir, err = DetectPath(os.Getenv("GOPATH"))
}

if err != nil {
	fmt.Println("Failed: ", err)
	return
}
```

File: main.go [hunk 3]
```diff
@@ -215,8 +232,39 @@ func ParseList(list *List, ref map[string]*List) (*ParsedList, error) {
...
+	return found
+}
+
+func isEnvSet(key string) bool {
+	if _, ok := os.LookupEnv(key); ok {
+		return true
+	}
+	return false
+}
+
```
Comment (iochen) [inline main.go:251]: unnecessary

File: main.go (hunk 2 above)
Comment (iochen) [reply main.go:266]:
```
var dir string
var err error
if *dataPath != "" {
	dir = *dataPath
} else {
	dir, err = DetectPath(os.Getenv("GOPATH"))
}
if err != nil {
	fmt.Println("Failed: ", err)
	return
}
```

the former one has something wrong

**Comment by Loyalsoldier** [comment]
Fix.

**Pull request merged**